### PR TITLE
Feature 22/user registration

### LINKED ISSUE #71: User registration
Description
As a User I want to register in cooperation system so I can use the system.

Preconditions:
Сo-owners of apartment building association shall be created in system.
Cooperation admin should send invitation e-mail with redirection link and register key due to user role.

Acceptance criteria

E-mail with redirection link shall redirect user to registration page.
When e-mail is already register in the system - system shall propose user to login.
When redirection link is clicked user redirect to the registration page with requires to input fields: First name, Middle name, Last name, e-mail, password, confirm password, registration key.
When all form fields filled in correctly button “Register” becomes available.
User can login system with e-mail and password with rights due to his role.
Fields fulfill requirements
Email address field: Max email length – 320 symbols, Min email length - 5 symbols, Recipient name shall consist of uppercase and lowercase letters in English (A-Z, a-z), digits from 0 to 9, special characters ! # $ % & ' * + - / = ? ^ _ ` { | ( ) , : ; < > @ [ \ ]
A special character cannot appear as the first or last character in an email address or appear consecutively two or more times.
Presence of one @ symbol is obligatory. Domain name shall consist of uppercase and lowercase letters in English (A-Z, a-z), digits from 0 to 9, hyphen (-), a period (.)
The address has to have a period (.) symbol placed after the domain name before the top top-level domain in an email address.

First name, Middle name, Last name: Max length – 50 symbols. Min length - 1 symbol. Symbols are uppercase and lowercase letters (A-Z, a-z). Special characters ! # $ % & ' * + - / = ? ^ _ ` { | ( ) , : ; < > @ [ \ ]

Password:
Max length – 128 symbols
Min length - 8 symbol
Symbols are uppercase and lowercase letters in English (A-Z, a-z), digits from 0 to 9

Password confirm: fields password and confirm password has to contain the same value.

Registration key (token): Length – 36 symbols. Symbols are uppercase and lowercase letters in English (A-Z, a-z), digits from 0 to 9

https://home-project-engineering.atlassian.net/browse/HP-13

Flow_registerUser

### PR DESCRIPTION
**Pull request by what-the-hub**
## Summary of issue

- created store
- rework validation
- add success and error toasts

### CODE REVIEW

File: src/views/RegisterUser.vue [hunk 1]
```diff
@@ -151,27 +151,68 @@ export default defineComponent({
...
+    async function sendInfo() {
+      const userData: UserStateInterface['Data'] = {
+        registration_token: state.formData.registrationKey,
+        first_name: state.formData.firstName,
+        last_name: state.formData.lastName,
+        email: state.formData.email,
+        password: state.formData.password.confirm,
+        contacts: [{ id: Date.now(), type: 'email', main: false, email: state.formData.email }],
+      };
+      store.dispatch('userStore/SET_USER_INFO', userData);
```
Comment (tsehla42) [inline src/views/RegisterUser.vue]: missing await for async fn call
vuex actions are async, composition API needs manual async handlers

File: src/views/RegisterUser.vue [hunk 2]
```diff
@@ -1,98 +1,91 @@
...
           @blur="v$.firstName.$touch"
         />
-        <small v-if="v$.firstName.$error" id="email-help" class="p-error">{{ v$.firstName.$errors[0].$message }}</small>
+        <small v-if="v$.firstName.$error" class="p-error">{{ v$.firstName.$errors[0].$message }}</small>
       </div>
       <div class="field">
         <input-text
           type="text"
-          v-model="state.midleName"
+          v-model="state.formData.midleName"
```
Comment (tsehla42) [inline src/views/RegisterUser.vue]: typo: mi**dd**leName

File: src/store/user/index.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import { UserStateInterface } from '@/store/user/types';
+import { getters } from '@/store/user/getters';
+import { mutations } from '@/store/user/mutations';
+import { actions } from '@/store/user/actions';
+
+export const state: UserStateInterface = {
+  Data: {
+    registration_token: '',
+    first_name: '',
+    // midle_name: '',
```
Comment (tsehla42) [inline src/store/user/index.ts]: typo: mi**dd**leName

File: src/store/user/index.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { Module } from 'vuex';
+import { RootStateInterface } from '@/store/types';
+import { UserStateInterface } from '@/store/user/types';
+import { getters } from '@/store/user/getters';
+import { mutations } from '@/store/user/mutations';
+import { actions } from '@/store/user/actions';
+
+export const state: UserStateInterface = {
+  Data: {
```
Comment (tsehla42) [inline src/store/user/index.ts]: naming conventions:
- Data => data
- registration_token => registrationToken

File: src/views/RegisterUser.vue (hunk 1 above)
Comment (tsehla42) [reply src/views/RegisterUser.vue]:
```js  
await store.dispatch('userStore/SET_USER_INFO', userData);
```

File: src/store/user/index.ts (hunk 4 above)
Comment (what-the-hub) [reply src/store/user/index.ts]: @asteof  i am agree with first edit but registration_token wil be used in future such as parameter in request body

Comment (tsehla42) [reply src/store/user/index.ts]: like this?
```js 
 [action]: ({ commit }, payload) => {
    HTTP.post(
      '/path', 
      { 
        registration_token: payload.registrationToken,
        first_name: payload.firstName,
       . . .
      })
      .then(() => { })
      .catch(() => { });
  },
```

Comment (what-the-hub) [reply src/store/user/index.ts]: @asteof  Yes but I was planning to send the state, without extra lines. I mean in the request body I can insert all the state **UserStateInterface.data**

**Review by tsehla42 — DISMISSED** [review]

File: src/views/RegisterUser.vue [hunk 5]
```diff
@@ -151,27 +151,68 @@ export default defineComponent({
...
-    const v$ = useVuelidate(rules, state);
+    const v$ = useVuelidate(rules, state.formData);
+    async function sendInfo() {
+      const userData: UserStateInterface['data'] = {
+        registrationToken: state.formData.registrationKey,
+        firstName: state.formData.firstName,
+        lastName: state.formData.lastName,
+        email: state.formData.email,
+        password: state.formData.password.confirm,
+        contacts: [{ id: Date.now(), type: 'email', main: false, email: state.formData.email }],
```
Comment (maveric-ui) [inline src/views/RegisterUser.vue]: is it necessary prop on step registration user?

File: src/views/RegisterUser.vue [hunk 6]
```diff
@@ -151,27 +151,68 @@ export default defineComponent({
...
+      state.formData.lastName = '';
+      state.formData.email = '';
+      state.formData.password = {
+        password: '',
+        confirm: '',
+      };
+      state.formData.registrationKey = '';
+    };
+
+    async function showStatus() {
```
Comment (maveric-ui) [inline src/views/RegisterUser.vue]: seems this method is used as submit, but better to divide and make more logical function calls

create method which called `onSubmit`- which will contain method `sendInfo` and `showStatus`

File: src/store/user/actions.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { ActionTree } from 'vuex';
+import { RootStateInterface } from '@/store/types';
+import { UserStateInterface, UserActionTypes, UserMutationTypes, Actions } from '@/store/user/types';
+import { HTTP } from '@/core/api/http-common';
+
+export const actions: ActionTree<UserStateInterface, RootStateInterface> & Actions = {
+  [UserActionTypes.SET_USER_INFO]: async ({ commit }, userData) => {
+    const bodyData = {
+      registration_token: userData?.registrationToken,
```
Comment (maveric-ui) [inline src/store/user/actions.ts]: why all props optional (undefined)
according to flow, userData have required props and can't be undefined

File: src/store/types.ts [hunk 8]
```diff
@@ -1,14 +1,15 @@
...
+import { UserStateInterface } from '@/store/user';
 import { Actions, Getters, Mutations } from '@/store/cooperation/types';
 
 export interface RootStateInterface {
   hello: string;
   world: string;
 }
 
-export type Store = CooperationStore<CooperationStateInterface>;
+export type Store = CooperationStore<CooperationStateInterface> & UserStore<UserStateInterface>;
```
Comment (maveric-ui) [inline src/store/types.ts]: maybe will be better
export interface StoreInterface {
  cooperationStore: CooperationStateInterface,
  userStore: UserStateInterface
}

File: src/views/RegisterUser.vue (hunk 5 above)
Comment (what-the-hub) [reply src/views/RegisterUser.vue]: yes. in post user we need this prop. But you are right, id didn`t necessar, its my mistake

**Review by markuwas — DISMISSED** [review]

**Review by tsehla42 — DISMISSED** [review]

**Review by tsehla42 — APPROVED** [review]

**Review by markuwas — APPROVED** [review]

**Pull request merged**